rclcpp/test/rclcpp/test_type_support.cpp

```diff
@@ -0,0 +1,196 @@
+      "topicTSParameterEvent",
+      *ts_parameter_event,
+      PublisherOptions()) {}
+};
+
+const rosidl_message_type_support_t *ts_set_parameter_result =
+  rclcpp::type_support::get_set_parameters_result_msg_type_support();
+
+class TestTSSetParameterResult : public rclcpp::PublisherBase
+{
+public:
+  explicit TestTSSetParameterResult(rclcpp::Node * node)
+  : rclcpp::PublisherBase(
+      node->get_node_base_interface().get(),
+      "topicTSSetParameterResult",
+      *ts_set_parameter_result,
+      PublisherOptions()) {}
+};
+
+const rosidl_message_type_support_t *ts_parameter_descriptor =
+  rclcpp::type_support::get_parameter_descriptor_msg_type_support();
+
+class TestTSParameterDescriptor : public rclcpp::PublisherBase
+{
+public:
+  explicit TestTSParameterDescriptor(rclcpp::Node * node)
+  : rclcpp::PublisherBase(
+      node->get_node_base_interface().get(),
+      "topicTSParameterDescriptor",
+      *ts_parameter_descriptor,
+      PublisherOptions()) {}
+};
+
+const rosidl_message_type_support_t *ts_list_parameter_result =
+  rclcpp::type_support::get_list_parameters_result_msg_type_support();
+
+class TestTSListParametersResult : public rclcpp::PublisherBase
+{
+public:
+  explicit TestTSListParametersResult(rclcpp::Node * node)
+  : rclcpp::PublisherBase(
+      node->get_node_base_interface().get(),
+      "topicTSListParametersResult",
+      *ts_list_parameter_result,
+      PublisherOptions()) {}
+};
+
+/*
+   Test that the publisher is created properly for different msg typesupport
+ */
+TEST_F(TestTypeSupport, basic_getters) {
+  initialize();
+  {
+    auto publisher = TestTSParameterEvent(node.get());
+    std::shared_ptr<const rcl_publisher_t> publisher_handle = publisher.get_publisher_handle();
+    EXPECT_NE(nullptr, publisher_handle);
+  }
+  {
+    auto publisher = TestTSSetParameterResult(node.get());
+    std::shared_ptr<const rcl_publisher_t> publisher_handle = publisher.get_publisher_handle();
+    EXPECT_NE(nullptr, publisher_handle);
+  }
+  {
+    auto publisher = TestTSParameterDescriptor(node.get());
+    std::shared_ptr<const rcl_publisher_t> publisher_handle = publisher.get_publisher_handle();
+    EXPECT_NE(nullptr, publisher_handle);
+  }
+  {
+    auto publisher = TestTSListParametersResult(node.get());
+    std::shared_ptr<const rcl_publisher_t> publisher_handle = publisher.get_publisher_handle();
+    EXPECT_NE(nullptr, publisher_handle);
+  }
+}
+
+void test_type_support_init_fini(const rosidl_service_type_support_t * ts){
+  auto node_handle_int = rclcpp::Node::make_shared("base_node");
+  rcl_service_t service_handle = rcl_get_zero_initialized_service();
+  rcl_service_options_t service_options = rcl_service_get_default_options();
+  rcl_ret_t ret = rcl_service_init(
+    &service_handle,
+    node_handle_int->get_node_base_interface()->get_rcl_node_handle(),
+    ts, "base_node_service", &service_options);
+  if (ret != RCL_RET_OK) {
+    FAIL();
+    return;
```

Comment:
Just for clarification here, `FAIL` is the equivalent of an `ASSERT` so you wouldn't need an early return, but ideally this is returning an `::testing::AssertionFailure` with a helpful message.

Reply:
811ad25